File: app/src/main/java/com/jgasteiz/readcomicsandroid/adapters/ItemListAdapter.java
package com.jgasteiz.readcomicsandroid.adapters;


import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.jgasteiz.readcomicsandroid.R;
import com.jgasteiz.readcomicsandroid.activities.BaseActivity;
import com.jgasteiz.readcomicsandroid.models.Item;
import com.jgasteiz.readcomicsandroid.models.ItemType;

import java.util.ArrayList;

public class ItemListAdapter extends RecyclerView.Adapter<ItemListAdapter.ViewHolder> {

    private ArrayList<Item> mItemList;
    private BaseActivity mContext;
    private View.OnClickListener mOnItemClick;
    private View.OnClickListener mOnDownloadClick;
    private View.OnClickListener mOnRemoveClick;

    public ItemListAdapter(BaseActivity context, ArrayList<Item> itemList, View.OnClickListener onItemClick, View.OnClickListener onDownloadClick, View.OnClickListener onRemoveClick) {
        mItemList = itemList;
        mContext = context;
        mOnItemClick = onItemClick;
        mOnDownloadClick = onDownloadClick;
        mOnRemoveClick = onRemoveClick;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item, parent, false);
        return new ViewHolder(view, mOnItemClick);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Item item = mItemList.get(position);
        holder.bindItem(item, mContext);

//        TODO
//        if (item.type == ItemType.COMIC) {
//            if (item.isComicDownloading) {
//                setDownloadInProgress(holder.itemView, item)
//            } else if (item.isComicOffline) {
//                setRemoveButton(holder.itemView, item)
//            } else {
//                setDownloadButton(holder.itemView, item)
//            }
//        }
    }

    @Override
    public int getItemCount() {
        return mItemList.size();
    }

    // TODO
    private void setDownloadButton() {

    }
    // TODO
    private void setDownloadInProgress() {

    }
    // TODO
    private void setRemoveButton() {

    }

    class ViewHolder extends RecyclerView.ViewHolder {
        private View.OnClickListener onItemClick;

        ViewHolder(View itemView, View.OnClickListener onItemClick) {
            super(itemView);
            this.onItemClick = onItemClick;
        }

        void bindItem(Item item, BaseActivity context) {
            ((TextView) itemView.findViewById(R.id.item_name)).setText(item.name);
            itemView.setOnClickListener(onItemClick);

            if (item.getType() == ItemType.COMIC) {
                itemView.findViewById(R.id.action_button).setVisibility(View.VISIBLE);
            } else {
                itemView.findViewById(R.id.action_button).setVisibility(View.GONE);
            }
        }
    }
}

File: app/src/main/java/com/jgasteiz/readcomicsandroid/adapters/ItemListAdapter.kt
package com.jgasteiz.readcomicsandroid.adapters

import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import com.jgasteiz.readcomicsandroid.R
import com.jgasteiz.readcomicsandroid.activities.BaseActivity
import com.jgasteiz.readcomicsandroid.activities.DirectoryActivity
import com.jgasteiz.readcomicsandroid.helpers.Utils
import com.jgasteiz.readcomicsandroid.models.Item
import com.jgasteiz.readcomicsandroid.models.ItemType
import kotlinx.android.synthetic.main.list_item.view.*
import java.util.*
import kotlin.collections.ArrayList


class ItemListAdapter(private val context: BaseActivity,
                      private val itemList: ArrayList<Item>,
                      private val itemClick: (Item) -> Unit) : RecyclerView.Adapter<ItemListAdapter.ViewHolder>()
{

    private val LOG_TAG = ItemListAdapter::class.java.simpleName

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.list_item, parent, false)
        return ViewHolder(view, itemClick)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bindItem(itemList[position], context)

        if (Utils.isComicOffline(context, itemList[position])) {
            setRemoveButton(holder.itemView, itemList[position])
        } else {
            setDownloadButton(holder.itemView, itemList[position])
        }
    }

    override fun getItemCount(): Int {
        return itemList.size
    }

    class ViewHolder(view: View, private val itemClick: (Item) -> Unit) : RecyclerView.ViewHolder(view) {

        fun bindItem(item: Item, context: BaseActivity) {
            with(item) {
                itemView.item_name.text = item.name
                itemView.setOnClickListener { itemClick(this) }

                if (item.type == ItemType.COMIC) {
                    itemView.action_button.visibility = View.VISIBLE
                } else {
                    itemView.action_button.visibility = View.GONE
                }
            }
        }
    }

    /**
     * Change the action button to `download` a comic.
     * @param cView View instance
     * *
     * @param comic Comic instance
     */
    private fun setDownloadButton(cView: View, comic: Item) {
        val downloadButton = cView.findViewById<Button>(R.id.action_button)
        downloadButton.text = context.getString(R.string.download_comic)

        // Download the comic when the download button is clicked.
        downloadButton.setOnClickListener {
            val downloadComicButton = cView.findViewById<Button>(R.id.action_button)
            val progressTextView = cView.findViewById<TextView>(R.id.progress_text)

            downloadComicButton.visibility = View.GONE
            progressTextView.visibility = View.VISIBLE
            progressTextView.setText(R.string.downloading_comic)

            context.mService?.downloadComic(comic)

            // Check the download status.
            checkActiveDownload(comic, cView, progressTextView, downloadComicButton)
        }
    }

    /**
     * Change the action button to `remove` a downloaded comic.
     * @param convertView View instance
     * *
     * @param comic Comic instance
     */
    private fun setRemoveButton(convertView: View, comic: Item) {
        val removeComicButton = convertView.findViewById<Button>(R.id.action_button)
        removeComicButton.text = context.getString(R.string.remove_download)

        // Remove the downloaded comic when the button is clicked.
        removeComicButton.setOnClickListener {
            Utils.removeComicDownload(context, comic)
            if (context.hasRemovableItems) {
                itemList.remove(comic)
                notifyDataSetChanged()
            } else {
                setDownloadButton(convertView, comic)
            }
        }
    }

    /**
     * Check whether a download is taking place for the given comic or not, and update its status.
     * TODO: do this in the background, it currently blocks the main thread.
     */
    private fun checkActiveDownload (comic: Item, convertView: View, progressTextView: TextView, downloadComicButton: Button) {
        if (Utils.downloads[comic.path] == null) {
            return
        }

        val timer = Timer("checkDownloadProgress")

        timer.schedule(object : TimerTask() {
            override fun run() {
                val progress = Utils.downloads[comic.path] as Int
                if (progress < 100) {
                    Log.d(LOG_TAG, "Download progress: $progress%")
                    (context as DirectoryActivity).runOnUiThread {
                        val percentage = "$progress%"
                        progressTextView.text = percentage
                    }
                } else {
                    Log.d(LOG_TAG, "Comic downloaded")
                    (context as DirectoryActivity).runOnUiThread {
                        setRemoveButton(convertView, comic)
                        progressTextView.visibility = View.GONE
                        downloadComicButton.visibility = View.VISIBLE
                    }
                    timer.cancel()
                }
            }
        }, 0, 1000)
    }

}